File: back/src/main/scala/models/Handshake.scala
package chrilves.kuzh.back.models

import io.circe.*
import io.circe.syntax.*
import cats.syntax.eq.*
import cats.syntax.functor.*
import java.util.*
import chrilves.kuzh.back.lib.crypto.*
import chrilves.kuzh.back.*

object Handshake:
  enum Out:
    case Challenge(challenge: Array[Byte], identityProofNeeded: Boolean)
    case Error(error: String, fatal: Boolean)
    case Established(state: State[Any])

  object Out:
    given handshakeEncoder: Encoder[Handshake.Out] with
      final def apply(h: Handshake.Out): Json =
        h match
          case Challenge(challenge, identityProofNeeded) =>
            Json.obj(
              "tag"                   -> "challenge".asJson,
              "challenge"             -> Base64UrlEncoded.encode(challenge).asString.asJson,
              "identity_proof_needed" -> identityProofNeeded.asJson
            )
          case Error(error, fatal) =>
            Json.obj(
              "tag"   -> "error".asJson,
              "error" -> error.asJson,
              "fatal" -> fatal.asJson
            )
          case Established(state) =>
            Json.obj("tag" -> "established".asJson, "state" -> state.asJson)

  enum In:
    case Crententials(assembly: models.assembly.Info, member: Member.Fingerprint)
    case ChallengeResponse(signature: Signature[Array[Byte]], identityProof: Option[IdentityProof])

  object In:
    extension (cr: ChallengeResponse)
      def check(
          member: Member.Fingerprint,
          storedIdentityProof: Option[IdentityProof],
          challenge: Array[Byte]
      ): Option[IdentityProof] =
        // Validity + Cohenrence
        if (
          cr.identityProof.map(_.isValid).getOrElse(true) &&
          cr.identityProof
            .flatMap(id1 => storedIdentityProof.map(id2 => id1 === id2))
            .getOrElse(true)
        )
        then
          storedIdentityProof.orElse(cr.identityProof).flatMap { id =>
            withVerify[Option[IdentityProof]](id.verify.toECPublicKey) { f =>
              if f[Array[Byte]](Signed(challenge, cr.signature))
              then Some(id)
              else None
            }
          }
        else None

    given handshakeDecoder: Decoder[Handshake.In] with
      def apply(c: HCursor): Decoder.Result[Handshake.In] =
        import Decoder.resultInstance.*
        c.downField("tag").as[String].flatMap {
          case "credentials" =>
            for
              assembly <- c.downField("assembly").as[assembly.Info]
              member   <- c.downField("member").as[Member.Fingerprint]
            yield Crententials(assembly, member)

          case "challenge_response" =>
            for
              sig <- c.downField("signature").as[Signature[Array[Byte]]]
              id  <- c.downField("identity_proof").as[Option[IdentityProof]]
            yield ChallengeResponse(sig, id)

          case tag =>
            raiseError(DecodingFailure(s"Unknown handshake tag '${tag}' in ${c}", Nil))
        }

File: front/src/model/Handshake.ts
import { Base64URL } from "../lib/Base64URL";
import { AssemblyInfo } from "./assembly/AssembyInfo";
import { State } from "./assembly/State";
import { Membership, Fingerprint, Serial } from "./Crypto";

export namespace Handshake {
  export namespace In {
    export type Challenge = {
      readonly tag: "challenge";
      readonly challenge: string;
      readonly identity_proof_needed: boolean;
    };

    export type Error = {
      readonly tag: "error";
      readonly error: string;
      readonly fatal: boolean;
    };

    export type Established = {
      readonly tag: "established";
      readonly state: State;
    };
  }

  export type In = In.Challenge | In.Error | In.Established;

  export namespace Out {
    export type Crententials = {
      readonly tag: "credentials";
      readonly assembly: AssemblyInfo;
      readonly member: Fingerprint;
    };

    export function credentials(
      assembly: AssemblyInfo,
      member: Fingerprint
    ): Crententials {
      return {
        tag: "credentials",
        assembly: assembly,
        member: member,
      };
    }

    export type ChallengeResponse = {
      readonly tag: "challenge_response";
      readonly signature: string;
      readonly identity_proof: Serial.IdentityProof | null;
    };

    export function challengeResponse(
      signature: string,
      identityProof: Serial.IdentityProof | null
    ): ChallengeResponse {
      return {
        tag: "challenge_response",
        signature: signature,
        identity_proof: identityProof,
      };
    }
  }

  export type Out = Out.Crententials | Out.ChallengeResponse;

  export async function replyToChallenge(
    membership: Membership,
    challenge: In.Challenge
  ): Promise<Out.ChallengeResponse> {
    const signature = await membership.me.signB64(
      Base64URL.getInstance().decode(challenge.challenge)
    );
    let identityProof: Serial.IdentityProof | null;

    if (challenge.identity_proof_needed) {
      const ip = await membership.me.identityProof();
      identityProof = await ip.toJson();
    } else {
      identityProof = null;
    }

    return Out.challengeResponse(signature, identityProof);
  }
}
